File: tizen/js/pip-overlay.js
/* PipOverlay — picture-in-picture overlay layer for the Tizen player (#109 MVP).
 *
 * Renders an image or web (iframe) overlay into a #pip element that sits ABOVE the
 * playlist #stage. The playlist renderer (PlaylistPlayer / ZoneRenderer) NEVER touches
 * #pip, so showing/clearing an overlay cannot change what's playing underneath.
 *
 * MVP semantics:
 *  - single overlay slot, last-show-wins (a new show replaces the current one);
 *  - duration timer in seconds (0 = until explicitly cleared);
 *  - device:pip-clear (matching pip_id, or none) or the timer tears it down;
 *  - teardown is wrapped so a malformed payload can never wedge the layer.
 *
 * Orientation: app.js applies the SAME orientation transform to #pip as to #stage, so a
 * corner position ("top-right") tracks the top-right of the visible CONTENT in every
 * orientation. This module only positions the box WITHIN #pip's (already-oriented) box.
 *
 * Deferred (not MVP): video/rtsp overlay types, priority/stacking, close-button focus.
 */
function PipOverlay(pipEl, opts) {
  opts = opts || {};
  this.pip = pipEl;
  this.doc = opts.document || (typeof document !== 'undefined' ? document : null);
  this.log = (typeof opts.log === 'function') ? opts.log : function () {};
  // Injectable timers keep teardown deterministic under test; default to the globals.
  this._setTimeout = opts.setTimeout || (typeof setTimeout !== 'undefined' ? setTimeout : null);
  this._clearTimeout = opts.clearTimeout || (typeof clearTimeout !== 'undefined' ? clearTimeout : null);
  this.timer = null;
  this.current = null; // pip_id of the overlay currently showing (null when empty)
}

// Corner/center -> inline style offsets. 4% inset keeps the box off the bezel edge.
PipOverlay.POSITIONS = {
  'top-left': { top: '4%', left: '4%' },
  'top-right': { top: '4%', right: '4%' },
  'bottom-left': { bottom: '4%', left: '4%' },
  'bottom-right': { bottom: '4%', right: '4%' },
  'center': { top: '50%', left: '50%', transform: 'translate(-50%, -50%)' }
};

PipOverlay.prototype.show = function (p) {
  if (!p || !this.pip || !this.doc) return;
  try {
    this.teardown();               // single slot, last-show-wins
    var box = this._buildBox(p);
    this.pip.appendChild(box);
    this.current = p.pip_id || '(anon)';
    var dur = Number(p.duration);
    if (this._setTimeout && isFinite(dur) && dur > 0) {
      var self = this;
      this.timer = this._setTimeout(function () { self.clear(self.current); }, dur * 1000);
    }
    this.log('info', 'pip show ' + (p.type || '?') + ' ' + (p.pip_id || '') +
      ' pos=' + (p.position || 'top-right') + ' dur=' + (isFinite(dur) ? dur : 0));
  } catch (e) {
    // A malformed payload must never wedge the layer: tear down and stay usable.
    this.teardown();
    this.log('warn', 'pip show failed: ' + (e && e.message ? e.message : e));
  }
};

PipOverlay.prototype.clear = function (pipId) {
  // A clear carrying a pip_id only clears if it matches the showing overlay (so a stale
  // clear for a replaced overlay is a no-op); an omitted pip_id clears whatever shows.
  if (pipId && this.current && pipId !== this.current) return;
  var had = !!this.current;
  this.teardown();
  if (had) this.log('info', 'pip cleared' + (pipId ? ' ' + pipId : ''));
};

PipOverlay.prototype.teardown = function () {
  try { if (this.timer && this._clearTimeout) { this._clearTimeout(this.timer); } } catch (e) {}
  this.timer = null;
  this.current = null;
  try { if (this.pip) this.pip.innerHTML = ''; } catch (e) {}
};

PipOverlay.prototype._buildBox = function (p) {
  var d = this.doc;
  var box = d.createElement('div');
  var s = box.style;
  s.position = 'absolute';
  s.width = pipPx(p.width, 480);
  s.height = pipPx(p.height, 360);
  s.overflow = 'hidden';
  s.boxSizing = 'border-box';
  s.background = pipColor(p.background_color) || '#000000';
  s.zIndex = '2';
  if (p.opacity != null && isFinite(Number(p.opacity))) s.opacity = String(pipClamp(Number(p.opacity), 0, 1));
  if (p.border_radius != null && isFinite(Number(p.border_radius))) s.borderRadius = pipPx(p.border_radius, 0);

  var pos = PipOverlay.POSITIONS[p.position] || PipOverlay.POSITIONS['top-right'];
  for (var k in pos) { if (pos.hasOwnProperty(k)) s[k] = pos[k]; }

  var hasTitle = p.title != null && p.title !== '';
  if (hasTitle) {
    var bar = d.createElement('div');
    bar.textContent = String(p.title);
    var bs = bar.style;
    bs.font = '600 16px sans-serif';
    bs.padding = '6px 10px';
    bs.color = pipColor(p.title_color) || '#ffffff';
    bs.background = 'rgba(0,0,0,0.45)';
    bs.whiteSpace = 'nowrap';
    bs.overflow = 'hidden';
    bs.textOverflow = 'ellipsis';
    box.appendChild(bar);
  }

  var media;
  if (p.type === 'web') {
    media = d.createElement('iframe');
    media.setAttribute('frameborder', '0');
    media.setAttribute('scrolling', 'no');
    // Mute web audio by default: an empty allow= denies autoplay (incl. audio).
    media.setAttribute('allow', '');
    media.src = p.uri;
  } else { // 'image' (and any non-web MVP type defaults to image render)
    media = d.createElement('img');
    media.src = p.uri;
  }
  var ms = media.style;
  ms.display = 'block';
  ms.border = '0';
  ms.width = '100%';
  ms.height = hasTitle ? 'calc(100% - 32px)' : '100%';
  ms.objectFit = 'cover';
  box.appendChild(media);
  return box;
};

function pipPx(v, def) { var n = Number(v); if (!isFinite(n) || n <= 0) n = def; return n + 'px'; }
function pipClamp(n, lo, hi) { return n < lo ? lo : (n > hi ? hi : n); }
function pipColor(c) { return (typeof c === 'string' && /^#[0-9A-Fa-f]{6}$/.test(c)) ? c : null; }

if (typeof module !== 'undefined' && module.exports) module.exports = { PipOverlay: PipOverlay };

File: android/app/src/main/java/com/remotedisplay/player/player/PipOverlay.kt
package com.remotedisplay.player.player

import android.content.Context
import android.graphics.Color
import android.graphics.drawable.GradientDrawable
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.Gravity
import android.view.View
import android.webkit.WebView
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView
import com.remotedisplay.player.util.ImageLoader
import com.remotedisplay.player.util.WebViewSupport
import org.json.JSONObject

/**
 * PiP overlay layer (#109). Renders an image or web (WebView) overlay into [pipLayout],
 * which is the top child of rootLayout — so it draws above the playlist AND inherits the
 * orientation rotation/translation applied to rootView (corner positions track the visible
 * content). The playlist renderers never touch pipLayout.
 *
 * MVP semantics (mirrors the web/Tizen players): single overlay slot, last-show-wins;
 * duration timer (0 = until cleared); device:pip-clear (id-aware) or the timer tears it
 * down; teardown is wrapped so a malformed payload can't wedge the layer. Reports show/clear
 * via [log] (device:log tag "pip").
 *
 * All methods must run on the main thread (WebSocketService posts the socket events there).
 */
class PipOverlay(
    private val context: Context,
    private val pipLayout: FrameLayout,
    private val log: (level: String, message: String) -> Unit = { _, _ -> }
) {
    private val handler = Handler(Looper.getMainLooper())
    private var timer: Runnable? = null
    private var current: String? = null
    private var webView: WebView? = null

    fun show(p: JSONObject) {
        try {
            teardown() // single slot, last-show-wins
            val type = p.optString("type", "image")
            val uri = p.optString("uri", "")
            if (uri.isEmpty()) { log("warn", "pip show ignored: empty uri"); return }

            val dm = context.resources.displayMetrics
            val w = p.optInt("width", 480).coerceIn(1, dm.widthPixels * 4)
            val h = p.optInt("height", 360).coerceIn(1, dm.heightPixels * 4)

            val box = LinearLayout(context).apply {
                orientation = LinearLayout.VERTICAL
                clipToOutline = true
                val radius = p.optInt("border_radius", 0).toFloat()
                background = GradientDrawable().apply {
                    setColor(parseColor(p.optString("background_color", ""), Color.BLACK))
                    cornerRadius = radius
                }
                alpha = p.optDouble("opacity", 1.0).toFloat().coerceIn(0f, 1f)
            }

            val title = p.optString("title", "")
            if (title.isNotEmpty()) {
                box.addView(TextView(context).apply {
                    text = title
                    setTextColor(parseColor(p.optString("title_color", ""), Color.WHITE))
                    setBackgroundColor(Color.argb(115, 0, 0, 0))
                    textSize = 14f
                    maxLines = 1
                    setPadding(20, 12, 20, 12)
                }, LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT))
            }

            // Media fills the remaining box height (weight 1).
            val mediaLp = LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, 0, 1f)
            if (type == "web") {
                val wv = WebView(context)
                WebViewSupport.configure(wv, "PiP")
                // Mute web audio by default: deny autoplay-with-gesture so audio can't start.
                try { wv.settings.mediaPlaybackRequiresUserGesture = true } catch (_: Throwable) {}
                wv.setBackgroundColor(Color.TRANSPARENT)
                wv.loadUrl(uri)
                webView = wv
                box.addView(wv, mediaLp)
            } else {
                val img = ImageView(context).apply { scaleType = ImageView.ScaleType.CENTER_CROP }
                box.addView(img, mediaLp)
                loadImageInto(img, uri, w, h)
            }

            // Corner/center placement; 4% inset off the edges (matches web/Tizen).
            val lp = FrameLayout.LayoutParams(w, h)
            val mx = (dm.widthPixels * 0.04f).toInt()
            val my = (dm.heightPixels * 0.04f).toInt()
            lp.gravity = when (p.optString("position", "top-right")) {
                "top-left" -> { lp.leftMargin = mx; lp.topMargin = my; Gravity.TOP or Gravity.START }
                "bottom-right" -> { lp.rightMargin = mx; lp.bottomMargin = my; Gravity.BOTTOM or Gravity.END }
                "bottom-left" -> { lp.leftMargin = mx; lp.bottomMargin = my; Gravity.BOTTOM or Gravity.START }
                "center" -> Gravity.CENTER
                else -> { lp.rightMargin = mx; lp.topMargin = my; Gravity.TOP or Gravity.END } // top-right
            }

            pipLayout.addView(box, lp)
            pipLayout.visibility = View.VISIBLE
            current = p.optString("pip_id", "(anon)")

            val dur = p.optInt("duration", 0)
            if (dur > 0) {
                val id = current
                timer = Runnable { clear(id) }.also { handler.postDelayed(it, dur * 1000L) }
            }
            log("info", "pip show $type ${p.optString("pip_id", "")} pos=${p.optString("position", "top-right")} dur=$dur")
        } catch (e: Throwable) {
            // A malformed payload must never wedge the layer.
            teardown()
            log("warn", "pip show failed: ${e.message}")
        }
    }

    /** Clear a showing overlay. A pip_id only clears if it matches; an empty id clears any. */
    fun clear(pipId: String?) {
        if (!pipId.isNullOrEmpty() && current != null && pipId != current) return
        val had = current != null
        teardown()
        if (had) log("info", "pip cleared" + (if (!pipId.isNullOrEmpty()) " $pipId" else ""))
    }

    /** Convenience for the device:pip-clear payload ({ pip_id? }). */
    fun clearFrom(data: JSONObject) = clear(if (data.has("pip_id")) data.optString("pip_id") else null)

    private fun teardown() {
        try { timer?.let { handler.removeCallbacks(it) } } catch (_: Throwable) {}
        timer = null
        current = null
        try { webView?.apply { stopLoading(); loadUrl("about:blank"); destroy() } } catch (_: Throwable) {}
        webView = null
        try { pipLayout.removeAllViews(); pipLayout.visibility = View.GONE } catch (_: Throwable) {}
    }

    private fun loadImageInto(img: ImageView, url: String, w: Int, h: Int) {
        val token = current
        Thread {
            val bmp = try { ImageLoader.decodeUrl(url, w, h) } catch (e: Throwable) { null }
            img.post {
                // Drop the result if this overlay was torn down / replaced while decoding.
                if (img.parent == null || token != current) return@post
                if (bmp != null) img.setImageBitmap(bmp)
                else { log("warn", "pip image failed to load"); clear(token) }
            }
        }.start()
    }

    private fun parseColor(hex: String, fallback: Int): Int =
        if (hex.matches(Regex("^#[0-9A-Fa-f]{6}$"))) try { Color.parseColor(hex) } catch (e: Throwable) { fallback } else fallback
}
